frame/composable-traits/src/staking/mod.rs

```diff
@@ -1,35 +1,138 @@
+	/// pool.
+	pub total_rewards: Balance,
+
+	/// A book keeping field to track the actual total reward without the
+	/// reward dilution adjustment caused by new stakers joining the pool.
+	pub total_dilution_adjustment: Balance,
+
+	/// Upper bound on the `total_rewards - total_dilution_adjustment`.
+	pub max_rewards: Balance,
+
+	/// The rewarding rate that increases the pool `total_reward` (and `actual_total_reward`)
+	/// at a given time.
+	pub reward_rate: Perbill,
 }
 
-/// staking typed fNFT, usually can be mapped to raw fNFT storage type
-#[derive(Debug, PartialEq, Eq, Copy, Clone, Encode, Decode, TypeInfo)]
-pub struct Stake<Balance, Rewards> {
+impl<AssetId, Balance: Zero> Reward<AssetId, Balance> {
+	pub fn from(reward_config: RewardConfig<AssetId, Balance>) -> Reward<AssetId, Balance> {
+		Reward {
+			asset_id: reward_config.asset_id,
+			total_rewards: Zero::zero(),
+			total_dilution_adjustment: Zero::zero(),
+			max_rewards: reward_config.max_rewards,
+			reward_rate: reward_config.reward_rate,
+		}
+	}
+}
+
+/// A reward pool is a collection of rewards that are allocated to stakers to incentivize a
+/// particular purpose. Eg: a pool of rewards for incentivizing adding liquidity to a pablo swap
+/// pool. TODO refer to the relevant section in the design doc.
+#[derive(RuntimeDebug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub struct RewardPool<
+	AccountId,
+	AssetId,
+	Balance,
+	BlockNumber,
+	DurationPresets,
+	RewardsLength: Get<u32>,
+> {
+	pub owner: AccountId,
+
+	/// The staked asset id of the reward pool.
+	pub asset_id: AssetId,
+
+	/// rewards accumulated
+	pub rewards: BoundedBTreeMap<AssetId, Reward<AssetId, Balance>, RewardsLength>,
+
+	/// Total shares distributed among stakers
+	pub total_shares: Balance,
+
+	/// Pool would stop adding rewards to pool at this block number.
+	pub end_block: BlockNumber,
+
+	// possible lock config for this pool
+	pub lock: LockConfig<DurationPresets>,
+}
+
+/// Reward configurations for a given asset type.
+#[derive(RuntimeDebug, PartialEq, Eq, Clone, MaxEncodedLen, Encode, Decode, TypeInfo)]
+pub struct RewardConfig<AssetId, Balance> {
+	/// asset id of the reward
+	pub asset_id: AssetId,
+
+	/// Upper bound on the `total_rewards - total_dilution_adjustment`.
+	pub max_rewards: Balance,
+
+	/// The rewarding rate that increases the pool `total_reward` (and `actual_total_reward`)
+	/// at a given time.
+	pub reward_rate: Perbill,
+}
+
+/// Categorize the reward pool by it's incentive characteristics and expose
+/// initial configuration parameters.
+/// TODO refer to the relevant section in the design doc.
+#[derive(RuntimeDebug, Encode, Decode, MaxEncodedLen, Clone, PartialEq, Eq, TypeInfo)]
+pub enum RewardPoolConfiguration<AccountId, AssetId, Balance, BlockNumber, DurationPresets> {
+	/// A pool with an adjustable reward rate to be used as incentive.
+	RewardRateBasedIncentive {
```

Comment:
you mix things here. There will be extrinsic to add reward manually and also set rate to 0. So you traying to build non generic API. 

not sure I see multiplier for share in config

Reply:
Setting reward rate = 0 is a valid state. Any pool would support manual increase of rewards. Not a problem

Reply:
> /// Pool would stop adding rewards to pool at this block number.
> end_block: BlockNumber,
setting end_block to now also will end rate based rewarding at same time.

so there are 2 variables which can be set to  ensure it is no RewardRateBasedIncetive, not cleary why then it is called such?

Reply:
It is natural to set the reward rate to zero because we would no longer want to incentivize a certain pool after launching and running for a while. As I explained above it's part of how the incentive works, it can be turned off when owner/governance decides to do so.

Reply:
enum Incetives {
  RewardRateBasedIncentive(SharedConfig)
}

enum(struct) == sturct(enum) => you produce something not useful

Reply:
enum(struct) with the #[non_exhaustive] attribute is fine to communicate possible future changes. It should be zero overhead anyways.